src/Npgsql/Internal/TypeHandlers/TextHandler.cs

```diff
@@ -270,5 +271,13 @@ public Task Write(byte[] value, NpgsqlWriteBuffer buf, NpgsqlLengthCache? length
     #endregion
 
     /// <inheritdoc />
-    public virtual TextReader GetTextReader(Stream stream) => new StreamReader(stream, _encoding);
+    public virtual TextReader GetTextReader(Stream stream, int byteLength, NpgsqlReadBuffer buffer)
+    {
+        if (byteLength > 0 && buffer.ReadBytesLeft >= byteLength)
+        {
+            return new PreparedTextReader(_encoding.GetString(buffer.Buffer, buffer.ReadPosition, byteLength), (NpgsqlReadBuffer.ColumnStream)stream);
```

Comment:
In theory we could have a TextReader implementation which directly accepts a `ReadOnlySpan<byte>` and an encoding, this way we don't have to decode and have the entire string in-memory up-front. But that's probably too much for this API.

Reply:
Also, we could even have a single cached copy of PreparedTextReader and ColumnStream, since most usages never need more than one at any given point.

Reply:
> In theory we could have a TextReader implementation which directly accepts a ReadOnlySpan<byte> and an encoding, this way we don't have to decode and have the entire string in-memory up-front. But that's probably too much for this API.


Are you sure?
Wouldn't that only mean wrapping [`Encoder.Convert()`](https://source.dot.net/#System.Private.CoreLib/Encoder.cs,305) inside [`PreparedTextReader.Read()`](https://github.com/npgsql/npgsql/blob/6242ddbc205f9fe1ce78028ce356dc47a0437112/src/Npgsql/PreparedTextReader.cs#L49) and save allocating the whole string?

Reply:
> On the other hand this might even help the case where the string is not in the buffer and by that be worthwhile as a potential follow-up PR - no?

Re-implementing `StreamReader` just to reduce allocations seems to me a bit too far, but we can take a look at this at some later point, yeah.

Reply:
Yeah I agree.
Whenever I use StreamReader in a transient way, wrapping existing buffers, I take offense in it's extensive internal buffering but that's probably more of a mental problem on my part than a real life issue.

Reply:
> Whenever I use StreamReader in a transient way, wrapping existing buffers, I take offense in it's extensive internal buffering

That's actually the main reason for this pr :trollface:. In my use case it's the most allocating object.